kernel/comps/device/src/id.rs

```diff
@@ -0,0 +1,284 @@
+/// identifies the device driver, while the minor number identifies a specific device
+/// instance controlled by that driver.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct DeviceId {
+    major: u32,
+    minor: u32,
+}
+
+impl DeviceId {
+    fn new(major: u32, minor: u32) -> Self {
+        Self { major, minor }
+    }
+
+    /// Returns the major device number.
+    pub fn major(&self) -> u32 {
+        self.major
+    }
+
+    /// Returns the minor device number.
+    pub fn minor(&self) -> u32 {
+        self.minor
+    }
+}
+
+impl DeviceId {
+    /// Creates a device ID from the encoded `u64` value.
+    ///
+    /// See [`as_encoded_u64`] for details about how to encode a device ID to a `u64` value.
+    ///
+    /// [`as_encoded_u64`]: Self::as_encoded_u64
+    pub fn from_encoded_u64(raw: u64) -> Self {
+        let major = ((raw >> 32) & 0xffff_f000 | (raw >> 8) & 0x0000_0fff) as u32;
+        let minor = ((raw >> 12) & 0xffff_ff00 | raw & 0x0000_00ff) as u32;
+        Self::new(major, minor)
+    }
+
+    /// Encodes the device ID as a `u64` value.
+    ///
+    /// The lower 32 bits use the same encoding strategy as Linux. See the Linux implementation at:
+    /// <https://github.com/torvalds/linux/blob/0ff41df1cb268fc69e703a08a57ee14ae967d0ca/include/linux/kdev_t.h#L39-L44>.
+    ///
+    /// If the major or minor device number is too large, the additional bits will be recorded
+    /// using the higher 32 bits. Note that as of 2025, the Linux kernel still has no support for
+    /// 64-bit device IDs:
+    /// <https://github.com/torvalds/linux/blob/0ff41df1cb268fc69e703a08a57ee14ae967d0ca/include/linux/types.h#L18>.
+    /// So this encoding follows the implementation in glibc:
+    /// <https://github.com/bminor/glibc/blob/632d895f3e5d98162f77b9c3c1da4ec19968b671/bits/sysmacros.h#L26-L34>.
+    pub fn as_encoded_u64(&self) -> u64 {
+        let major = self.major() as u64;
+        let minor = self.minor() as u64;
+        ((major & 0xffff_f000) << 32)
+            | ((major & 0x0000_0fff) << 8)
+            | ((minor & 0xffff_ff00) << 12)
+            | (minor & 0x0000_00ff)
+    }
+}
+
+/// An allocator for device IDs.
+///
+/// This structure manages the allocation and release of device IDs for a specific
+/// device type, major number, and range of minor numbers. It ensures that each
+/// minor number within the specified range is allocated at most once.
+///
+/// When the allocator is dropped, it automatically unregisters the device IDs
+/// with the appropriate subsystem (block or character device management).
+#[derive(Debug)]
+pub struct DeviceIdAllocator {
+    /// The type of device this allocator manages IDs for.
+    pub type_: DeviceType,
+    /// The major device number managed by this allocator.
+    pub major: u32,
+    /// The range of minor device numbers managed by this allocator.
+    pub minors: Range<u32>,
+    used: RwLock<BTreeSet<u32>>,
```

Comment:
Consider using `ostd::util::IdSet`, which is superior than `BTreeSet` for managing a fixed-size set of 256 items.